services/web/server/src/simcore_service_webserver/projects/projects_api.py

```diff
@@ -477,32 +528,213 @@ async def trigger_connected_service_retrieve(
+        log.debug(
+            "removing project [%s] from user [%s] resources", project_uuid, user_id
+        )
+        await rt.remove(PROJECT_ID_KEY)
+    # check it is not opened by someone else
+    user_to_session_ids.remove((user_id, client_session_id))
+    log.debug("remaining user_to_session_ids: %s", user_to_session_ids)
+    if not user_to_session_ids:
+        # NOTE: depending on the garbage collector speed, it might already be removing it
+        fire_and_forget_task(
+            remove_project_interactive_services(user_id, project_uuid, app)
+        )
+    else:
+        log.warning(
+            "project [%s] is used by other users: [%s]. This should not be possible",
+            project_uuid,
+            {uid for uid, _ in user_to_session_ids},
+        )
+
+
 async def _get_project_lock_state(
-    user_id: int, project_uuid: str, app: web.Application
+    user_id: int,
+    project_uuid: str,
+    app: web.Application,
 ) -> ProjectLocked:
+    """returns the lock state of a project
+    1. If a project is locked for any reason, first return the project as locked and STATUS defined by lock
+    2. If a client_session_id is passed, then first check to see if the project is currently opened by this very user/tab combination, if yes returns the project as Locked and OPENED.
+    3. If any other user that user_id is using the project (even disconnected before the TTL is finished) then the project is Locked and OPENED.
+    4. If the same user is using the project with a valid socket id (meaning a tab is currently active) then the project is Locked and OPENED.
+    5. If the same user is using the project with NO socket id (meaning there is no current tab active) then the project is Unlocked and OPENED. which means the user can open it again.
+    """
+    log.debug("getting project [%s] lock state for user [%s]...", project_uuid, user_id)
+    prj_locked_state: Optional[ProjectLocked] = await get_project_locked_state(
+        app, project_uuid
+    )
+    if prj_locked_state:
+        return prj_locked_state
+
+    # let's now check if anyone has the project in use somehow
     with managed_resource(user_id, None, app) as rt:
-        # checks who is using it
-        users_of_project = await rt.find_users_of_resource("project_id", project_uuid)
-        usernames = [await get_user_name(app, uid) for uid in set(users_of_project)]
-        assert (
-            len(usernames) <= 1
-        )  # nosec  # currently not possible to have more than 1
-
-        # based on usage, sets an state
-        is_locked: bool = len(usernames) > 0
-        if is_locked:
+        user_session_id_list: List[Tuple[int, str]] = await rt.find_users_of_resource(
+            PROJECT_ID_KEY, project_uuid
+        )
+    set_user_ids = {x for x, _ in user_session_id_list}
+
+    assert (
+        len(set_user_ids) <= 1
+    )  # nosec  # NOTE: A project can only be opened by one user in one tab at the moment
```

Comment:
```suggestion
    )  # nosec  # NOTE: A project can only be opened by one user in one tab
```

Reply:
this is not meant to remain like this forever. we should allow that later.